### PR TITLE
Remove FlutterUndoManagerPlugin handlers from undo manager on dealloc

### PR DESCRIPTION
**Pull request by jmagman**
When `FlutterUndoManagerPlugin` deallocated, it is supposed to deregister itself from the `NSUndoManager` stack.
However, during dealloc `_undoManagerDelegate.undoManager` is nil (`_undoManagerDelegate` = the engine, and `undoManager` was its view controller's `undoManager`, which was already gone).

Since `_undoManagerDelegate.undoManager` is nil, it doesn't actually call `-[NSUndoManager removeAllActionsWithTarget]`. In the add-to-app scenario, after the view controller pops back to the native view, and "undo" is invoked, the undo action for the `FlutterUndoManagerPlugin` is handled, but it already dealloced so crash.

Fixes https://github.com/flutter/flutter/issues/150408

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterUndoManagerPluginTest.mm [hunk 1]
```diff
@@ -164,4 +165,38 @@ - (void)testSetUndoStateDoesInteractWithInputDelegate {
...
+    XCTAssertTrue(undoManager.canUndo);
+    XCTAssertTrue(undoManager.canRedo);
+    // Fake out the undoManager being nil, which happens when the FlutterViewController deallocs and
+    // the undo manager can't be fetched from the FlutterEngine delegate.
+    undoManagerDelegate.undoManager = nil;
+  }
+  XCTAssertNil(weakUndoManagerPlugin);
+  // Regression test for https://github.com/flutter/flutter/issues/150408.
+  // Undo manager undo and redo stack should be empty after the plugin deallocs.
+  XCTAssertFalse(undoManager.canUndo);
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterUndoManagerPluginTest.mm:198]: Wrote a new test that fails on main without this change (though I couldn't get it to actually crash):
```
FlutterUndoManagerPluginTest.mm:198: error: -[FlutterUndoManagerPluginTest testDeallocRemovesAllUndoManagerActions] : ((undoManager.canUndo) is false) failed
```

**Review by hellohuanlin — APPROVED** [review]
tricky one!

**Comment by dullgrass** [comment]
@jmagman I'm a bit confused. According to the https://github.com/flutter/flutter/issues/150408 description, if we set `undomanagerPlugin` to nil when it's released, there won't be any crashes. Why not do it this way?

**Comment by jmagman** [comment]
> @jmagman I'm a bit confused. According to the [flutter/flutter#150408](https://github.com/flutter/flutter/issues/150408) description, if we set `undomanagerPlugin` to nil when it's released, there won't be any crashes. Why not do it this way?

I don't understand your question, does https://github.com/flutter/flutter/issues/150408#issuecomment-2188104799 answer it?

**Pull request merged**